pysgrid/sgrid.py

```diff
@@ -252,6 +267,287 @@ def _save_common_components(self, nc_file):
                     dataset_grid_var.axes = ' '.join(axes)
         return grid_vars
 
+    def build_edges(self):
+        '''
+        TODO: This should produce something that would allow a drawing library to graph all the lines
+        '''
+        pass
+
+    def locate_faces(self, points):
+        """
+        Returns the face indices, one per point.
+
+        Points that are not in the mesh will have an index of -1
+
+        If a single point is passed in, a single index will be returned
+        If a sequence of points is passed in an array of indexes will be returned.
+
+        :param points:  The points that you want to locate -- (lon, lat). If the shape of point
+                        is 1D, function will return a scalar index. If it is 2D, it will return
+                        a 1D array of indices
+        :type points: array-like containing one or more points: shape (2,) for one point, shape (N, 2)
+                     for more than one point.
+
+        This version utilizes the CellTree data structure.
+
+        """
+        points = np.asarray(points, dtype=np.float64)
+        just_one = (points.ndim == 1)
+        points.shape = (-1, 2)
+
+        try:
+            import cell_tree2d
+        except ImportError:
+            raise ImportError("the cell_tree2d package must be installed to use the celltree search:\n"
+                              "https://github.com/NOAA-ORR-ERD/cell_tree2d/")
+        if not hasattr(self, '_tree') or self._tree is None:
+            self.build_celltree()
+        indices = self._tree.multi_locate(points)
+        node_x = indices % (self.node_lat.shape[1] - 1)
+        node_y = indices // (self.node_lat.shape[1] - 1)
+        node_ind = np.column_stack((node_y, node_x))
+        if just_one:
+            return node_ind[0]
+        else:
+            return node_ind
+
+    def get_variable_by_index(self, var, index):
+        '''
+        Function to get the node values of a given face index.
+        Emulates the self.grid.nodes[self.grid.nodes.faces[index]] paradigm of unstructured grids
+        '''
+
+        arr = var[:]
+        x = index[:, 0]
+        y = index[:, 1]
+        return np.ma.column_stack((arr[x, y], arr[x + 1, y], arr[x + 1, y + 1], arr[x, y + 1]))
+
+    def build_celltree(self):
+        """
+        Tries to build the celltree for the current UGrid. Will fail if nodes
+        or faces is not defined.
```

Comment:
What do you mean by not defined?  Should they always be defined of the dataset is non-compliant?